scripts/globals/pets/avatar.lua

```diff
@@ -0,0 +1,385 @@
+
+            if
+                petArg and
+                (abilityName == 'assault' or
+                abilityName == 'retreat')
+            then
+                -- reset cast cooldown buffer and timer
+                local lastCastTime = pet:getLocalVar(lastCastTimeVar)
+                local lastCastTimeStamp = pet:getLocalVar(lastCastTimeStampVar)
+                if
+                    lastCastTimeStamp > 0
+                then
+                    lastCastTime = lastCastTime + os.time() - lastCastTimeStamp
+                end
+
+                pet:setLocalVar(lastCastTimeStampVar, 0)
+                pet:setLocalVar(lastCastTimeVar, lastCastTime)
+                pet:setLocalVar(buffModeVar, 1)
+                setMagicCastCooldown(petArg)
+            end
+        end)
+    end
+end
+
+xi.pets.avatar.onMobDeath = function(pet)
+    local master = pet:getMaster()
+
+    if master and master:getObjType() == xi.objType.PC then
+        master:removeListener(playerListenerVar)
+        master:removeListener(playerListenerVar .. 'ABILITY')
+    end
+end
+
+xi.pets.avatar.onMobMagicPrepare = function(pet)
+    -- Note that:
+    -- returning -1 in this function forces TryCastSpell to exit without casting a spell, but
+    -- will still set the m_LastMagicTime to ensure next call of this function is after the cast delay
+    -- Also, if we return nothing TryCastSpell will default to normal mob casting behavior (nukes from spell list, etc)
+-- print('onMobMagicPrepare: ' .. pet:getMobMod(xi.mobMod.MAGIC_COOL))
+    local master = pet:getMaster()
+
+    if master and master:getObjType() == xi.objType.PC then
+        -- meta checks for fresh pet, etc
+        pet:setLocalVar(lastCastTimeVar, 0)
+        pet:setLocalVar(lastCastTimeStampVar, os.time())
+        local spellID = 0
+        local spellTarget = nil
+        if pet:getMobMod(xi.mobMod.MAGIC_COOL) == 1 then
+            setMagicCastCooldown(pet)
+
+            return -1
+        end
+
+        pet:setLocalVar(buffModeVar, 1)
+
+        -- Core functionality to decide which spell to use
+        spellID, spellTarget = xi.pets.avatar.getSpiritSpell(pet)
+
+        -- Final items to cast the spell and ensure cast delay is proper
+        local spell = GetSpell(spellID)
+        if spell then
+            if spell:getSkillType() == xi.skill.ENHANCING_MAGIC then
+                -- half casting delay
+                pet:setLocalVar(buffModeVar, 0)
+            end
+-- print(spellTarget:getName() .. '->' .. spellID)
+            pet:castSpell(spellID, spellTarget or pet)
+            setMagicCastCooldown(pet)
+            return -1
+        end
+    end
+end
+
+xi.pets.avatar.getSpiritSpell = function(pet)
+    local spellID = 0
+    local spellTarget = nil
+    local petID = pet:getPetID()
+    if petID == xi.petId.LIGHT_SPIRIT then
+        spellID, spellTarget = xi.pets.avatar.getLightSpiritSpell(pet)
+    -- elseif not pet:isEngaged() then
```

Comment:
If the commented-out code isn't needed, remove it?

Reply:
Good point, done